macaroonbakery/codec.py

```diff
@@ -65,22 +70,91 @@ def _encode_caveat_v1(condition, root_key, third_party_pub_key, key):
             Base64Encoder).decode('ascii'),
         'Nonce': base64.b64encode(nonce).decode('ascii'),
         'Id': base64.b64encode(encrypted).decode('ascii')
-    }))).decode('ascii')
+    })))
+
+
+PUBLIC_KEY_PREFIX_LEN = 4
+KEY_LEN = 32
+
+
+def _encode_caveat_v2_v3(version, condition, root_key, third_party_pub_key,
+                         key, ns):
+    '''Create a version 2 or version 3 third-party caveat.
+
+    The format has the following packed binary fields (note
+    that all fields up to and including the nonce are the same
+    as the v2 format):
+
+        version 2 or 3 [1 byte]
+        first 4 bytes of third-party Curve25519 public key [4 bytes]
+        first-party Curve25519 public key [32 bytes]
+        nonce [24 bytes]
+        encrypted secret part [rest of message]
+
+    The encrypted part encrypts the following fields
+    with box.Seal:
+
+        version 2 or 3 [1 byte]
+        length of root key [n: uvarint]
+        root key [n bytes]
+        length of encoded namespace [n: uvarint] (Version 3 only)
+        encoded namespace [n bytes] (Version 3 only)
+        condition [rest of encrypted part]
+    '''
+    ns_data = b''
+    if version >= bakery.BAKERY_V3:
+        ns_data = ns.serialize()
+    data = six.int2byte(version)
+    data = data + third_party_pub_key.encode()[:PUBLIC_KEY_PREFIX_LEN]
+    data = data + key.public_key.encode()[:]
+    secret = _encode_secret_part_v2_v3(version, condition, root_key, ns_data)
+    box = Box(key, third_party_pub_key)
+    encrypted = box.encrypt(secret)
+    nonce = encrypted[0:Box.NONCE_SIZE]
+    encrypted = encrypted[Box.NONCE_SIZE:]
+    data = data + nonce[:]
+    data = data + encrypted
+    return data
+
+
+def _encode_secret_part_v2_v3(version, condition, root_key, ns):
+    '''Creates a version 2 or version 3 secret part of the third party
+    caveat. The returned data is not encrypted.
+
+    The format has the following packed binary fields:
+    version 2 or 3 [1 byte]
+    root key length [n: uvarint]
+    root key [n bytes]
+    namespace length [n: uvarint] (v3 only)
+    namespace [n bytes] (v3 only)
+    predicate [rest of message]
+    '''
+    data = six.int2byte(version)
+    data = _encode_varint(len(root_key), data)
+    data = data + root_key
+    if version >= bakery.BAKERY_V3:
+        data = _encode_varint(len(ns), data)
+        data = data + ns
+    data = data + condition.encode('utf-8')
+    return data
 
 
 def decode_caveat(key, caveat):
     '''Decode caveat by decrypting the encrypted part using key.
 
     @param key the nacl private key to decode.
-    @param caveat a base64 encoded JSON string.
+    @param caveat bytes.
     @return ThirdPartyCaveatInfo
     '''
     if len(caveat) == 0:
         raise ValueError('empty third party caveat')
 
-    first = caveat[0]
-    if first == 'e':
+    first = caveat[:1]
+    if first == b'e':
         return _decode_caveat_v1(key, caveat)
```

Comment:
I'd suggest retaining the comment from the original Go code here:

    # 'e' will be the first byte if the caveatid is a base64 encoded JSON object.